File: Naive Bayes.py
#import packages
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer,TfidfTransformer
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder

# Loading the data set
salary_train = pd.read_csv("SalaryData_Train.csv",encoding = "ISO-8859-1")
salary_test = pd.read_csv("SalaryData_Test.csv",encoding = "ISO-8859-1")

#combining training and testing data to process the data
salary_data = salary_train.append(salary_test)

#removing irrelevant information
salary_data.drop(["education","relationship","capitalgain","capitalloss"], axis = 1, inplace = True)

#converting into binary
lb=LabelEncoder()
salary_data["age"]=lb.fit_transform(salary_data["age"])
salary_data["workclass"]=lb.fit_transform(salary_data["workclass"])
salary_data["maritalstatus"]=lb.fit_transform(salary_data["maritalstatus"])
salary_data["occupation"]=lb.fit_transform(salary_data["occupation"])
salary_data["race"]=lb.fit_transform(salary_data["race"])
salary_data["sex"]=lb.fit_transform(salary_data["sex"])
salary_data["hoursperweek"]=lb.fit_transform(salary_data["hoursperweek"])
salary_data["native"]=lb.fit_transform(salary_data["native"])
salary_data["Salary"]=lb.fit_transform(salary_data["Salary"])
salary_data.describe()
salary_data

# Normalization function 
def norm_func(i):
    x = (i-i.min())	/ (i.max()-i.min())
    return (x)

# Normalized data frame
x = norm_func(salary_data.iloc[:, :9])
x.describe()

#target
y = salary_data.iloc[:, 9:]
y

# splitting data into train and test data sets 
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 9)

# Preparing a naive bayes model on training data set 
from sklearn.naive_bayes import MultinomialNB as MB

# Multinomial Naive Bayes
classifier_mb = MB()

#for train data
classifier_mb.fit(x_train, y_train)
from sklearn.metrics import accuracy_score
y_pred = classifier_mb.predict(x_train) # store the prediction data
accuracy_score(y_train,y_pred) # calculate the accuracy

#for test data
classifier_mb.fit(x_test, y_test)
y_pred = classifier_mb.predict(x_test) 
accuracy_score(y_test,y_pred)

File: Naive Bayes.R
# Import the salary dataset
library(readr)
library(plyr)
library(ROCR)

salary_train <- read_csv("SalaryData_Train.csv")
salary_test <- read_csv("SalaryData_Test.csv")
salary <- rbind(salary_train, salary_test)

View(salary)
str(salary)
attach(salary)

salary$Salary=factor(ifelse(salary$Salary==">50K","1","0"))

#removing data without relative information
salary$education=NULL
salary$relationship=NULL
salary$capitalgain=NULL
salary$capitalloss=NULL

colnames(salary)
col=c("occupation","race","sex","native")
salary[col] <- lapply(salary[col], factor)

salary$age=factor(cut(age,breaks=4,labels = FALSE))
salary$hoursperweek=factor(cut(hoursperweek,breaks = 4,labels = FALSE))
str(salary)

#combining to reduce levels
salary$workclass <- factor(gsub('^Federal-gov', 'Government', salary$workclass))
salary$workclass <- factor(gsub('^Local-gov', 'Government', salary$workclass))
salary$workclass <- factor(gsub('^State-gov', 'Government', salary$workclass))
salary$workclass=factor(gsub('^Self-emp-inc','self-emp',salary$workclass))
salary$workclass=factor(gsub('^Self-emp-not-inc','self-emp',salary$workclass))
str(salary$workclass)

salary$maritalstatus=factor(gsub('^Married-AF-spouse','Married',salary$maritalstatus))
salary$maritalstatus=factor(gsub('^Married-civ-spouse','Married',salary$maritalstatus))
salary$maritalstatus=factor(gsub('^Married-spouse-absent','Married',salary$maritalstatus))
salary$maritalstatus=factor(gsub('^Divorced','Single',salary$maritalstatus))
salary$maritalstatus=factor(gsub('^Separated','Single',salary$maritalstatus))
salary$maritalstatus=factor(gsub('^Widowed','Single',salary$maritalstatus))
salary$maritalstatus=factor(gsub('^Never-married','Single',salary$maritalstatus))
str(salary$maritalstatus)

str(salary$occupation)
levels(salary$occupation)
salary$occupation=factor(gsub('^Prof-specialty','Proffesional',salary$occupation))
salary$occupation=factor(gsub('^Craft-repair','Workers',salary$occupation))
salary$occupation=factor(gsub('^Exec-managerial','Administration',salary$occupation))
salary$occupation=factor(gsub('^Adm-clerical','Administration',salary$occupation))
salary$occupation=factor(gsub('^Other-service','Services',salary$occupation))
salary$occupation=factor(gsub('^Armed-Forces','Services',salary$occupation))
salary$occupation=factor(gsub('^Farming-fishing','Workers',salary$occupation))
salary$occupation=factor(gsub('^Handlers-cleaners','Workers',salary$occupation))
salary$occupation=factor(gsub('^Machine-op-inspct','Workers',salary$occupation))
salary$occupation=factor(gsub('^Priv-house-serv','Services',salary$occupation))
salary$occupation=factor(gsub('^Protective-serv','Services',salary$occupation))
salary$occupation=factor(gsub('^Tech-support','Workers',salary$occupation))
salary$occupation=factor(gsub('^Transport-moving','Workers',salary$occupation))

summary(salary$occupation)
salary$native=NULL

summary(salary)

#Partitioning the data into training and testing:
id=sample(2,nrow(salary),prob = c(0.8,0.2),replace = TRUE)
train=salary[id==1,]
test=salary[id==2,]

#Generating the model:
library(e1071)
#install.packages("caret")
library(caret)
model=naiveBayes(train, train$Salary)
model_pred=predict(model,test)
confusionMatrix(test$Salary,model_pred)

library(gmodels)
CrossTable(model_pred, test$Salary,
           prop.chisq = FALSE, prop.t = FALSE, prop.r = FALSE,
           dnn = c('predicted', 'actual'))

test_acc = mean(model_pred == test$Salary)
test_acc

# On Training Data
train_pred <- predict(model, train)

train_acc = mean(train_pred == train$Salary)
train_acc